### PR TITLE
Add ability for clients to register hooks for device state endpoint

### PR DESCRIPTION
**Pull request by dfunckt**
Change-type: minor

### CODE REVIEW

File: src/lib/device-state.ts [hunk 1]
```diff
@@ -97,6 +97,30 @@ export const serviceInstallFromImage = (
...
+
+export function registerHook<
+	K extends keyof HookRegistry,
+	Fn extends typeof STATE_ENDPOINT_HOOKS[K][number]
+>(name: K, fn: Fn) {
+	STATE_ENDPOINT_HOOKS[name].push(fn);
+}
+
+function runHook(name: keyof HookRegistry, args: any[]) {
+	STATE_ENDPOINT_HOOKS[name].forEach(hook => hook.apply(null, args));
```
Comment (Page-) [inline src/lib/device-state.ts]:
```suggestion
	STATE_ENDPOINT_HOOKS[name].forEach(hook => hook(...args));
```

File: src/lib/device-state.ts [hunk 2]
```diff
@@ -97,6 +97,30 @@ export const serviceInstallFromImage = (
...
+};
+
+export function registerHook<
+	K extends keyof HookRegistry,
+	Fn extends typeof STATE_ENDPOINT_HOOKS[K][number]
+>(name: K, fn: Fn) {
+	STATE_ENDPOINT_HOOKS[name].push(fn);
+}
+
+function runHook(name: keyof HookRegistry, args: any[]) {
```
Comment (Page-) [inline src/lib/device-state.ts]: You should be able to get strict types with
```suggestion
function runHook<T extends keyof HookRegistry>(name: T, args: Parameters<HookRegistry[T][number]>) {
```

Comment (dfunckt) [reply src/lib/device-state.ts]: Ah nice, I searched for something like `Parameters` but didn't find it.

File: src/lib/device-state.ts (hunk 1 above)
Comment (dfunckt) [reply src/lib/device-state.ts]: This unfortunately doesn't work, even after adapting the signature as you suggested below.

Comment (Page-) [reply src/lib/device-state.ts]: You can do
```suggestion
	STATE_ENDPOINT_HOOKS[name].forEach(hook => hook(...args as Parameters<typeof hook>));
```
instead and still be able to use `...` instead

File: src/lib/device-state.ts [hunk 3]
```diff
@@ -97,6 +97,33 @@ export const serviceInstallFromImage = (
...
+export function registerHook<
+	Key extends keyof HookRegistry,
+	Fn extends typeof STATE_ENDPOINT_HOOKS[Key][number]
+>(name: Key, fn: Fn) {
+	STATE_ENDPOINT_HOOKS[name].push(fn);
+}
+
+function runHook<
+	Key extends keyof HookRegistry,
+	Args extends Parameters<typeof STATE_ENDPOINT_HOOKS[Key][number]>
```
Comment (Page-) [inline src/lib/device-state.ts]: This allows changing what `Args` is, which shouldn't be possible as it should always be precisely `Parameters<typeof STATE_ENDPOINT_HOOKS[Key][number]>`, hence why I used `args: Parameters<HookRegistry[T][number]>`

File: src/lib/device-state.ts [hunk 4]
```diff
@@ -97,6 +97,33 @@ export const serviceInstallFromImage = (
...
+	deviceConfig: DeviceConfigHook[];
+}
+
+const STATE_ENDPOINT_HOOKS: HookRegistry = {
+	deviceConfig: [],
+};
+
+export function registerHook<
+	Key extends keyof HookRegistry,
+	Fn extends typeof STATE_ENDPOINT_HOOKS[Key][number]
```
Comment (Page-) [inline src/lib/device-state.ts]: Same here, this isn't really a generic and should instead be `fn: typeof STATE_ENDPOINT_HOOKS[Key][number]`

File: src/lib/device-state.ts [hunk 5]
```diff
@@ -47,6 +47,35 @@ export const serviceInstallFromImage = (
...
+) {
+	STATE_ENDPOINT_HOOKS[name].push(fn);
+}
+
+function runHook<Key extends keyof HookRegistry>(
+	name: Key,
+	args: Parameters<typeof STATE_ENDPOINT_HOOKS[Key][number]>,
+) {
+	STATE_ENDPOINT_HOOKS[name].forEach(hook =>
+		hook(...(args as Parameters<typeof hook>)),
```
Comment (Page-) [inline src/lib/device-state.ts]: Is this `as` actually necessary?

Comment (dfunckt) [reply src/lib/device-state.ts]: It doesn't work without this. It was your suggestion actually, my initial implementation was doing `hook.apply` which would loose type info but didn't type coercion.

Comment (Page-) [reply src/lib/device-state.ts]: Yeah, I do prefer the explicit cast because it at least makes it clear, but wanted to check

**Pull request merged**